N/A: Unicons Implementation

ivanahuckova opened the pull request:
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

ivanahuckova committed:
Create Unicon component

ivanahuckova committed:
Up;date import

ivanahuckova committed:
Migrate Icons to Unicons if replacement icon exists

ivanahuckova committed:
Monochrome icons implementation

ivanahuckova commented on packages/grafana-ui/src/components/Icon/Icon.tsx:
```diff
@@ -1,11 +1,18 @@
-import React from 'react';
```
@dprokop @torkelo Could you look at this new Icon implementation and let me know what do you think?

I have decided to currently import **react-unicons package** and whenever I replace previous icon with the new one, I add it to the icon types. This way, we will have a list of all icons and we can then decide wether we want to keep  **react-unicons package**  or wether we want to just import the icons that we are going to use. 

I have added 1 monochrome icon, to test out that. implementation works for both options.

ivanahuckova replied:
Also, do we do [tree shaking](https://webpack.js.org/guides/tree-shaking/) for grafana/ui? I suspect that no, because I didn't find it in webpack, but maybe I wasn't looking at the right place. Just a thing that I am considering would be useful,  if we decided to use react-unicons package.

ivanahuckova committed:
Update Icon.tsx

ivanahuckova committed:
Update Icon.tsx

torkelo commented on packages/grafana-ui/src/components/Forms/Input/Input.tsx:
```diff
@@ -246,7 +245,7 @@ export const Input = React.forwardRef<HTMLInputElement, Props>((props, ref) => {
         {(suffix || loading) && (
           <div className={styles.suffix} ref={suffixRef}>
-            {loading && <Icon name="spinner" className={cx('fa-spin', styles.loadingIndicator)} />}
+            {loading && <i className={cx('fa fa-spinner', 'fa-spin', styles.loadingIndicator)} />}
```
this looks strange :)

ivanahuckova replied:
:D Temporary fix,  as we don't have spinner icon.  Also I have converted Icons that have specific styling to <i /> to make sure that we are not losing any styling. I will definitely refactor it to <Icon /> when we agree on how </Icon > is implemented.

torkelo commented on packages/grafana-ui/src/components/Icon/Icon.tsx:
```diff
@@ -1,11 +1,18 @@
 import { IconType } from './types';
 
-export interface IconProps {
+type Size = 'xs' | 'sm' | 'base' | 'md' | 'lg';
```
not sure about the 'base'. think we should use same type as button size: https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/Button/types.ts#L5 

Think we should move to grafana-ui/types and name it ComponentSize so we can share that type between components

ivanahuckova replied:
Yeah, that is a good idea! I'll do that then.

torkelo commented on packages/grafana-ui/src/components/Icon/Icon.tsx:
```diff
@@ -1,11 +1,18 @@
-export interface IconProps {
+type Size = 'xs' | 'sm' | 'base' | 'md' | 'lg';
+
+interface IconProps extends Themeable {
```
wonder if can use something like  make it extend HtmlElementAttributes or something so you do not need to add className or onClick onMouseDown etc, 

Also maybe consider adding React.forwardRef

dprokop replied:
yes, agree on this

torkelo reviewed (commented):
Looks like a really good start, think the Icon component you have created is the in the right direction !

torkelo commented on packages/grafana-ui/src/components/Icon/assets/exclamation-triangle.ts:
```diff
@@ -0,0 +1,34 @@
+    React.createElement('path', {
+      fill: color,
+      d: 'M12,14a1,1,0,0,1-1-1V9a1,1,0,0,1,2,0v4A1,1,0,0,1,12,14Z',
+    })
```
is this for the monochrome version? Did you have to do this manually or download it from unicons?

ivanahuckova replied:
I had to do this manually, but creating component as bellow might be more clear/understandable solution. We just always need to pass  `width: size`, `height: size`, color, secondaryColor to the svg component. Colors will be generated in Icon component. We need to generate 2 - one primary (darker) and then for lighter one (probbaly just adjust transparency, or create function that calculates what is the complimentary color) 

```import React, { FunctionComponent } from 'react';

type Props = {
  size: number;
  color: string;
  secondColor: string;
};

const Svg: FunctionComponent<Props> = ({ size, color, secondColor }) => {
  return (
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width={size} height={size}>
      <path
        fill={secondColor}
        d="M20.05713,22H3.94287A3.02288,3.02288,0,0,1,1.3252,17.46631L9.38232,3.51123a3.02272,3.02272,0,0,1,5.23536,0L22.6748,17.46631A3.02288,3.02288,0,0,1,20.05713,22Z"
      />
      <circle cx="12" cy="17" r="1" fill={color} />
      <path fill={color} d="M12,14a1,1,0,0,1-1-1V9a1,1,0,0,1,2,0v4A1,1,0,0,1,12,14Z" />
    </svg>
  );
};

export default Svg;```

ivanahuckova committed:
Update monochrome svg import file

dprokop commented on packages/grafana-ui/package.json:
```diff
@@ -31,6 +31,7 @@
     "@grafana/data": "7.0.0-pre.0",
     "@grafana/slate-react": "0.22.9-grafana",
     "@grafana/tsconfig": "^1.0.0-rc1",
+    "@iconscout/react-unicons": "^1.0.0",
```
```suggestion
    "@iconscout/react-unicons": "1.0.0",
```

dprokop commented on packages/grafana-ui/src/components/Icon/Icon.tsx:
```diff
@@ -1,33 +1,93 @@
-  className?: string;
+  size?: Size;
+  color?: string;
+  type?: 'icon' | 'monochrome';
```
Icon type icon feels weird ;) maybe just `'color' | 'mono'`?

torkelo replied:
type: 'mono' | 'default'

dprokop commented on packages/grafana-ui/src/components/Icon/Icon.tsx:
```diff
@@ -1,33 +1,93 @@
+  icon: null | ComponentType<SvgProps>;
+}
+
+class UnThemedIcon extends PureComponent<IconProps, IconState> {
```
Minor: can we refactor this to functional component to be able to use `useTheme` hook? `withTheme` introduces some unnecessary complexity to this simple component

dprokop reviewed (commented):
Good progress @ivanahuckova! left few nits to think about

ivanahuckova committed:
Update packages/grafana-ui/package.json

Co-Authored-By: Dominik Prokop <dominik.prokop@grafana.com>

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivana/unicons

ivanahuckova committed:
Update Icon.tsx

ivanahuckova committed:
Merge branch 'ivana/unicons' of https://github.com/grafana/grafana into ivana/unicons

ivanahuckova committed:
Refactor Icon to functional component

ivanahuckova committed:
Merge branch 'master' into ivana/unicons

ivanahuckova committed:
Update Icon

ivanahuckova committed:
Merge branch 'ivana/unicons' of https://github.com/grafana/grafana into ivana/unicons

ivanahuckova commented:
@torkelo @dprokop 
Icon component:
- Refactored to functional component, so useTheme can be used
- props type = 'mono' | 'default' (instead of icon and monochrome)
-  added React.forwardRef
- props extend HtmlElementAttributes
- instead of size, we use ComponentSize from ` '../../types/size';`

I am now moving to second phase - updating icons.

torkelo commented:
Great job!
